### ISSUE TITLE
Lead time > 0 and testing yields incorrect samples

### ISSUE THREAD

**Issue by raffienficiaud** [awslabs/gluon-ts#1936]
## Description
I believe the behaviour of `TestSplitSampler` together with `InstanceSplitter` is incorrect when `lead_time` > 0 as it generates
* a `future_target` that is empty
* a `time_series_fields` that are empty
* a `forecast_start` outside of the testing set

## To Reproduce

Here is the creation of a simple dataset. There is only one additional sample on the test set vs. training set.
```python
df = pd.DataFrame(
    data={
        "value": np.arange(0, 21),
        "date": pd.date_range(
            pd.Timestamp.fromisoformat("2019-12-31"),
            pd.Timestamp.fromisoformat("2021-08-31"),
            freq="M",
        ),
    },
)
begin = df.date.min().replace(tzinfo=None)

for idx, value in enumerate((997, 1009, 1013, 1019, 1021, 1031), start=1):
    df.loc[df.index[-idx], "value"] = -value

all_datasets_prediction = {
    "target": df["value"].to_numpy(),
    "start": begin,
}

all_datasets_train = {
    "target": all_datasets_prediction["target"][:-1],
    "start": begin,
}

train_ds = ListDataset(
    [all_datasets_train] * 10,
    freq="M",
    one_dim_target=True,  # univariate for the moment
)

test_ds = ListDataset(
    [all_datasets_prediction] * 10,
    freq="M",
    one_dim_target=True,  # univariate for the moment
)
```

`df` contains this
```
    value       date
0       0 2019-12-31
1       1 2020-01-31
2       2 2020-02-29
3       3 2020-03-31
4       4 2020-04-30
5       5 2020-05-31
6       6 2020-06-30
7       7 2020-07-31
8       8 2020-08-31
9       9 2020-09-30
10     10 2020-10-31
11     11 2020-11-30
12     12 2020-12-31
13     13 2021-01-31
14     14 2021-02-28
15  -1031 2021-03-31
16  -1021 2021-04-30
17  -1019 2021-05-31
18  -1013 2021-06-30
19  -1009 2021-07-31
20   -997 2021-08-31      # this sample is not in the train set
```

I then generate the test samples loader like this, where `prediction_length=1`, `context_length=3` and `lead_time=1`:
```python
transforms = Chain(
    [
        AddConstFeature(
            output_field="myfield",
            pred_length=prediction_length,
            const=42,
            target_field=FieldName.TARGET,
        ),
        AddAgeFeature(
            target_field=FieldName.TARGET,
            output_field=FieldName.FEAT_AGE,
            pred_length=prediction_length,
            log_scale=False,
        ),
        VstackFeatures(
            output_field=FieldName.FEAT_TIME,
            input_fields=[
                "myfield",
                FieldName.FEAT_AGE,
            ],
        ),
    ]
)

instance_sampler = TestSplitSampler()

prediction_splitter = InstanceSplitter(
    target_field=FieldName.TARGET,
    is_pad_field=FieldName.IS_PAD,
    start_field=FieldName.START,
    forecast_start_field=FieldName.FORECAST_START,
    instance_sampler=instance_sampler,
    past_length=context_length,
    future_length=prediction_length,
    time_series_fields=[FieldName.FEAT_TIME],
    lead_time=lead_time,
)

data_loader = InferenceDataLoader(
    test_ds,
    batch_size=1,
    stack_fn=batchify,
    transform=transforms + prediction_splitter,
)
```

## Error message or code output

`print(first_ = next(iter(data_loader))` shows the following values:

* `forecast_start = [Timestamp('2021-10-31 00:00:00', freq='M')]`: this is outside the test dataset
* `future_target = tensor([], size=(1, 0))`: it is empty and the evaluation cannot actually evaluate anything
* `past_target = tensor([[-1013., -1009.,  -997.]])` it includes the sample we want to get a prediction for
* `future_time_feat = tensor([], size=(1, 0, 2))`: it is empty, while we should always be able to compute the age.

When I replace the `instance_sampler` above with 
```python
instance_sampler = PredictionSplitSampler(
    allow_empty_interval=False,
    axis=-1,
    min_past=0,
    min_future=lead_time + prediction_length, # --> this is the main difference with TestSplitSampler <--
)
```

I have the correct values for 

* `past_target = tensor([[-1021., -1019., -1013.]])`: the -1009 does not appear anywhere)
* `future_time_feat = tensor([[[42., 20.]]])` that is not empty  correct for August (and not empty)
* `future_target = tensor([[-997.]])` that is the ground truth for August
* there is one value of the sequence of prime numbers missing, which is correct with `lead_time=1`
* `forecast_start = [Timestamp('2021-08-31 00:00:00', freq='M')]`: actually the forecast should be starting on `2021-07-31` for getting a prediction for August (and not `2021-08-31`). This is my understanding of what a customer would see: I am able to make a prediction in July, on which I apply a lead time of 1 month for getting a value in August. But maybe this should be understood as "for when I get the actual values"?

## Environment
- Operating system: Ubuntu 20.04
- Python version: 3.8
- GluonTS version: master/
- MXNet version: not applicable

The example above involves only little of `pandas`, `torch` and `gluonts`.

This is one of the changes I have made in #1932 but I am not confident with it.

**Comment by raffienficiaud** [comment]
After digging for some time into this, it turns out that the work is done by the function `make_evaluation_predictions`: it strips the tail by `prediction_length + lead_time`, I guess this is to accommodate for the `TestSplitSampler`

**Comment by lostella** [comment]
@raffienficiaud thanks again for the detailed report! Let me try to address your points one by one:

> `future_target = tensor([], size=(1, 0))`: it is empty and the evaluation cannot actually evaluate anything

I believe this is the intended behavior. If anything, `TestSplitSampler` is a bit of a misnomer, since it suggests that all data required for testing the model (e.g. evaluating the loss or any other accuracy metric) will be there: it actually makes sure the data we're predicting does not enter the batch, so to avoid any potential source of leakage (like bugs); instead, evaluations can be done outside of the network (usually this is done using the `predictor` object that results from training, using the `Evaluator` class, but of course one is free to do it whichever way they prefer).

In fact, this is what backtesting utilities like `make_evaluation_prediction` or `backtest_metrics` do, specifically [here](https://github.com/awslabs/gluon-ts/blob/cf58988b53e545962b35631cc0fa943498461fbf/src/gluonts/evaluation/backtest.py#L94)

> `forecast_start = [Timestamp('2021-10-31 00:00:00', freq='M')]`: this is outside the test dataset

This also agrees with the protocol I'm describing above: in pseudocode, `forecast_start` is `start + freq * (lead_time + len(target))`, so it falls right next to the end of the series you are loading (which are from `test_ds`) when `lead_time=0`, and it's one timestep further if `lead_time=1`, and so on.

> `past_target = tensor([[-1013., -1009., -997.]])` it includes the sample we want to get a prediction for

Same here, it is assumed that all given data in the `target` field belongs to the past, i.e. predictions should be conditioned on all of the `target` data.

> `future_time_feat = tensor([], size=(1, 0, 2))`: it is empty, while we should always be able to compute the age.

This is the real issue. What the correct behavior should be is a bit of model dependent: should the `InstanceSplitter` output features from the whole `lead_time + prediction_length` range? Or only for the `prediction_length` range?
